File: python/src/apx_agent/_service_policies.py
"""Portable Service Policy declarations and deterministic local decisions.

The declaration is deliberately independent of a Databricks client. Native
deployment consumes the same validated models, while local execution uses the
ordered decision helper and existing policy primitives.
"""

from __future__ import annotations

from collections.abc import Callable, Sequence
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, StrictInt, StrictStr, field_validator, model_validator


class _StringEnum(str, Enum):
    """String enum that serializes to the public configuration value."""


class ServicePolicyTargetType(_StringEnum):
    MCP_SERVICE = "mcp_service"
    MODEL_SERVICE = "model_service"
    MODEL_PROVIDER_SERVICE = "model_provider_service"
    AGENT_SERVICE = "agent_service"


class ServicePolicyKind(_StringEnum):
    BUILTIN = "builtin"
    LLM_JUDGE = "llm_judge"
    SQL = "sql"


class BuiltinServicePolicy(_StringEnum):
    SENSITIVE_DATA = "sensitive_data"
    UNSAFE_CONTENT = "unsafe_content"
    JAILBREAK = "jailbreak"
    HALLUCINATION = "hallucination"


class ServicePolicyPhase(_StringEnum):
    ON_CALL = "on_call"
    ON_RESULT = "on_result"
    BOTH = "both"


class ServicePolicyMode(_StringEnum):
    ENFORCE = "enforce"
    DRY_RUN = "dry_run"


class LocalPolicyMode(_StringEnum):
    MIRROR = "mirror"
    OFF = "off"


class NativePolicyMode(_StringEnum):
    OFF = "off"
    PLAN = "plan"
    APPLY = "apply"
    REQUIRED = "required"


class ServicePolicyAction(_StringEnum):
    ALLOW = "ALLOW"
    ASK = "ASK"
    DENY = "DENY"
    UNAVAILABLE = "UNAVAILABLE"


class _StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ServicePolicy(_StrictModel):
    """One named policy definition within an attachment."""

    name: StrictStr
    kind: ServicePolicyKind
    builtin: BuiltinServicePolicy | None = None
    classifier: StrictStr | None = None
    prompt: StrictStr | None = None
    function: StrictStr | None = None
    phase: ServicePolicyPhase = ServicePolicyPhase.ON_CALL
    rank: StrictInt = Field(default=100, ge=0)

    @field_validator("name", "classifier", "prompt", "function")
    @classmethod
    def _non_empty(cls, value: str | None) -> str | None:
        if value is not None and not value.strip():
            raise ValueError("service policy string fields must not be empty")
        return value

    @model_validator(mode="after")
    def _validate_kind(self) -> ServicePolicy:
        references = {
            "builtin": self.builtin,
            "classifier": self.classifier,
            "prompt": self.prompt,
            "function": self.function,
        }
        if self.kind is ServicePolicyKind.BUILTIN:
            if self.builtin is None:
                raise ValueError("builtin policies require builtin")
            if self.builtin is BuiltinServicePolicy.JAILBREAK and self.phase is not ServicePolicyPhase.ON_CALL:
                raise ValueError("jailbreak policies are request/on_call only")
            if self.builtin is BuiltinServicePolicy.HALLUCINATION and self.phase is not ServicePolicyPhase.ON_RESULT:
                raise ValueError("hallucination policies are result/on_result only")
            if any(references[key] is not None for key in ("classifier", "prompt", "function")):
                raise ValueError("builtin policies cannot define classifier, prompt, or function")
        elif self.kind is ServicePolicyKind.LLM_JUDGE:
            if not self.classifier or not self.prompt:
                raise ValueError("llm_judge policies require classifier and prompt")
            if self.builtin is not None or self.function is not None:
                raise ValueError("llm_judge policies cannot define builtin or function")
        elif self.kind is ServicePolicyKind.SQL:
            if not self.function:
                raise ValueError("sql policies require a Unity Catalog function")
            if any(references[key] is not None for key in ("builtin", "classifier", "prompt")):
                raise ValueError("sql policies cannot define builtin, classifier, or prompt")
        return self


class ServicePolicyTarget(_StrictModel):
    """A reusable target identity for plan and diagnostic output."""

    target_type: ServicePolicyTargetType
    target: StrictStr

    @field_validator("target")
    @classmethod
    def _target_must_be_non_empty(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("service policy target must not be empty")
        return value


class ServicePolicyAttachment(_StrictModel):
    """Policies attached to one service target."""

    name: StrictStr
    target_type: ServicePolicyTargetType
    target: StrictStr
    mode: ServicePolicyMode = ServicePolicyMode.ENFORCE
    policies: list[ServicePolicy] = Field(default_factory=list)

    @field_validator("name", "target")
    @classmethod
    def _non_empty(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("service policy attachment strings must not be empty")
        return value


class AbacSelector(_StrictModel):
    """Future tag-based scope, retained as desired state until native support."""

    tags: dict[StrictStr, StrictStr] = Field(default_factory=dict)


class ServicePoliciesConfig(_StrictModel):
    """Declarative Service Policy configuration under ``AgentConfig``."""

    local_mode: LocalPolicyMode = LocalPolicyMode.MIRROR
    native_mode: NativePolicyMode = NativePolicyMode.OFF
    attachments: list[ServicePolicyAttachment] = Field(default_factory=list)
    abac: AbacSelector | None = None

    @model_validator(mode="after")
    def _validate_native_requirements(self) -> ServicePoliciesConfig:
        if self.native_mode is NativePolicyMode.REQUIRED and not self.attachments:
            raise ValueError("native_mode='required' needs at least one attachment")
        return self


@dataclass(frozen=True)
class ServicePolicyEvent:
    """Normalized event passed to a local policy evaluator."""

    phase: str
    target_type: ServicePolicyTargetType
    target: str
    content: str | None = None
    tool_name: str | None = None
    arguments: dict[str, Any] | None = None
    result: Any = None
    context: dict[str, Any] = field(default_factory=dict)


@dataclass(frozen=True)
class ServicePolicyDecision:
    """A sanitized policy decision shared by local and native adapters."""

    action: Literal["ALLOW", "DENY", "ASK", "UNAVAILABLE"]
    reason: str | None
    policy_name: str | None
    phase: str
    rank: int | None
    mode: ServicePolicyMode
    adapter: Literal["local", "native"]


@dataclass(frozen=True)
class _NormalizedAction:
    action: ServicePolicyAction
    reason: str | None


class ServicePolicyCapabilityError(ValueError):
    """Raised when a requested native policy cannot be represented."""


class ServicePolicyEvaluationError(ValueError):
    """Raised for malformed or unavailable policy evaluation inputs."""


ServicePolicyEvaluator = Callable[
    [ServicePolicy, ServicePolicyEvent],
    Any,
]


def ordered_policies(
    policies: Sequence[ServicePolicy],
    phase: ServicePolicyPhase | str,
) -> list[ServicePolicy]:
    """Return policies applicable to *phase* in deterministic rank order."""

    requested = ServicePolicyPhase(phase)
    if requested is ServicePolicyPhase.BOTH:
        applicable = [
            policy for policy in policies if policy.phase in (ServicePolicyPhase.ON_CALL, ServicePolicyPhase.ON_RESULT, ServicePolicyPhase.BOTH)
        ]
        reverse = False
    else:
        applicable = [
            policy for policy in policies if policy.phase in (requested, ServicePolicyPhase.BOTH)
        ]
        reverse = requested is ServicePolicyPhase.ON_RESULT
    return sorted(applicable, key=lambda policy: policy.rank, reverse=reverse)


def _normalize_action(value: Any) -> _NormalizedAction:
    """Normalize existing ``PolicyResult``/enum/string outputs."""

    reason = getattr(value, "reason", None)
    action = getattr(value, "action", value)
    if hasattr(action, "name"):
        action = action.name
    try:
        return _NormalizedAction(ServicePolicyAction(str(action).upper()), reason)
    except ValueError as exc:
        raise ServicePolicyEvaluationError(f"unknown policy action {action!r}") from exc


def evaluate_ordered_service_policies(
    policies: Sequence[ServicePolicy],
    event: ServicePolicyEvent,
    *,
    mode: ServicePolicyMode,
    adapter: Literal["local", "native"],
    evaluator: ServicePolicyEvaluator | None = None,
) -> ServicePolicyDecision:
    """Evaluate in rank order; stop at the first ``DENY`` or ``ASK``.

    ``dry_run`` changes enforcement at the adapter boundary, not the recorded
    action: operators must be able to see what would have been blocked.
    """

    phase = ServicePolicyPhase(event.phase)
    for policy in ordered_policies(policies, phase):
        try:
            result = evaluator(policy, event) if evaluator is not None else None
            if result is None:
                continue
            normalized = _normalize_action(result)
            action = normalized.action
            reason = normalized.reason
        except Exception as exc:
            reason = f"policy evaluation unavailable ({exc})"
            action = ServicePolicyAction.UNAVAILABLE if mode is ServicePolicyMode.DRY_RUN else ServicePolicyAction.DENY
        if action is ServicePolicyAction.ASK:
            target_type = ServicePolicyTargetType(event.target_type)
            if target_type is not ServicePolicyTargetType.MCP_SERVICE or phase is not ServicePolicyPhase.ON_CALL:
                reason = reason or "ASK is supported only for MCP on_call policies"
                action = ServicePolicyAction.UNAVAILABLE if mode is ServicePolicyMode.DRY_RUN else ServicePolicyAction.DENY
        if action is not ServicePolicyAction.ALLOW:
            return ServicePolicyDecision(
                action=action.value,
                reason=reason,
                policy_name=policy.name,
                phase=event.phase,
                rank=policy.rank,
                mode=mode,
                adapter=adapter,
            )
    return ServicePolicyDecision(
        action=ServicePolicyAction.ALLOW.value,
        reason=None,
        policy_name=None,
        phase=event.phase,
        rank=None,
        mode=mode,
        adapter=adapter,
    )

File: typescript/src/service-policies.ts
/** Portable Service Policy declarations and deterministic local decisions. */

export type ServicePolicyTargetType =
  | 'mcp_service'
  | 'model_service'
  | 'model_provider_service'
  | 'agent_service';
export type ServicePolicyKind = 'builtin' | 'llm_judge' | 'sql';
export type BuiltinServicePolicy = 'sensitive_data' | 'unsafe_content' | 'jailbreak' | 'hallucination';
export type ServicePolicyPhase = 'on_call' | 'on_result' | 'both';
export type ServicePolicyMode = 'enforce' | 'dry_run';
export type LocalPolicyMode = 'mirror' | 'off';
export type NativePolicyMode = 'off' | 'plan' | 'apply' | 'required';
export type ServicePolicyAction = 'ALLOW' | 'ASK' | 'DENY' | 'UNAVAILABLE';

export interface ServicePolicy {
  name: string;
  kind: ServicePolicyKind;
  builtin?: BuiltinServicePolicy;
  classifier?: string;
  prompt?: string;
  function?: string;
  phase?: ServicePolicyPhase;
  rank?: number;
}

export interface ServicePolicyAttachment {
  name: string;
  target_type: ServicePolicyTargetType;
  target: string;
  mode?: ServicePolicyMode;
  policies: ServicePolicy[];
}

export interface AbacSelector {
  tags: Record<string, string>;
}

export interface ServicePoliciesConfig {
  local_mode?: LocalPolicyMode;
  native_mode?: NativePolicyMode;
  attachments?: ServicePolicyAttachment[];
  abac?: AbacSelector;
}

export interface ServicePolicyEvent {
  phase: ServicePolicyPhase | string;
  target_type: ServicePolicyTargetType;
  target: string;
  content?: string;
  tool_name?: string;
  arguments?: Record<string, unknown>;
  result?: unknown;
  context?: Record<string, unknown>;
}

export interface ServicePolicyDecision {
  action: ServicePolicyAction;
  reason: string | null;
  policy_name: string | null;
  phase: string;
  rank: number | null;
  mode: ServicePolicyMode;
  adapter: 'local' | 'native';
}

export type ServicePolicyEvaluationResult =
  | ServicePolicyAction
  | { action: ServicePolicyAction | string; reason?: string | null }
  | null
  | undefined;

export type ServicePolicyEvaluator = (
  policy: ServicePolicy,
  event: ServicePolicyEvent,
) => ServicePolicyEvaluationResult;

const TARGET_TYPES: ReadonlySet<string> = new Set([
  'mcp_service',
  'model_service',
  'model_provider_service',
  'agent_service',
]);
const KINDS: ReadonlySet<string> = new Set(['builtin', 'llm_judge', 'sql']);
const BUILTINS: ReadonlySet<string> = new Set([
  'sensitive_data',
  'unsafe_content',
  'jailbreak',
  'hallucination',
]);
const PHASES: ReadonlySet<string> = new Set(['on_call', 'on_result', 'both']);
const MODES: ReadonlySet<string> = new Set(['enforce', 'dry_run']);
const LOCAL_MODES: ReadonlySet<string> = new Set(['mirror', 'off']);
const NATIVE_MODES: ReadonlySet<string> = new Set(['off', 'plan', 'apply', 'required']);

const keys = (value: Record<string, unknown>): Set<string> => new Set(Object.keys(value));

function assertObject(value: unknown, label: string): Record<string, unknown> {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    throw new TypeError(`${label} must be an object`);
  }
  return value as Record<string, unknown>;
}

function assertKeys(value: Record<string, unknown>, allowed: ReadonlySet<string>, label: string): void {
  for (const key of keys(value)) {
    if (!allowed.has(key)) throw new TypeError(`${label} has unknown field ${key}`);
  }
}

function stringValue(value: unknown, label: string): string {
  if (typeof value !== 'string' || value.trim() === '') throw new TypeError(`${label} must be non-empty`);
  return value;
}

function enumValue<T extends string>(value: unknown, allowed: ReadonlySet<string>, fallback: T, label: string): T {
  const selected = value ?? fallback;
  if (typeof selected !== 'string' || !allowed.has(selected)) throw new TypeError(`${label} is invalid`);
  return selected as T;
}

function normalizePolicy(value: unknown): ServicePolicy {
  const raw = assertObject(value, 'service policy');
  assertKeys(raw, new Set(['name', 'kind', 'builtin', 'classifier', 'prompt', 'function', 'phase', 'rank']), 'service policy');
  if (raw.kind === undefined) throw new TypeError('service policy kind is required');
  const kind = enumValue<ServicePolicyKind>(raw.kind, KINDS, 'builtin', 'service policy kind');
  const phase = enumValue<ServicePolicyPhase>(raw.phase, PHASES, 'on_call', 'service policy phase');
  const rank = raw.rank ?? 100;
  if (typeof rank !== 'number' || !Number.isInteger(rank) || rank < 0) throw new TypeError('service policy rank must be a non-negative integer');
  const policy: ServicePolicy = { name: stringValue(raw.name, 'service policy name'), kind, phase, rank };

  if (kind === 'builtin') {
    if (raw.builtin === undefined) throw new TypeError('builtin policies require builtin');
    policy.builtin = enumValue<BuiltinServicePolicy>(raw.builtin, BUILTINS, 'sensitive_data', 'builtin policy');
    if (policy.builtin === 'jailbreak' && phase !== 'on_call') throw new TypeError('jailbreak policies are request/on_call only');
    if (policy.builtin === 'hallucination' && phase !== 'on_result') throw new TypeError('hallucination policies are result/on_result only');
    if (raw.classifier !== undefined || raw.prompt !== undefined || raw.function !== undefined) throw new TypeError('builtin policies cannot define classifier, prompt, or function');
  } else if (kind === 'llm_judge') {
    policy.classifier = stringValue(raw.classifier, 'llm_judge classifier');
    policy.prompt = stringValue(raw.prompt, 'llm_judge prompt');
    if (raw.builtin !== undefined || raw.function !== undefined) throw new TypeError('llm_judge policies cannot define builtin or function');
  } else {
    policy.function = stringValue(raw.function, 'sql function');
    if (raw.builtin !== undefined || raw.classifier !== undefined || raw.prompt !== undefined) throw new TypeError('sql policies cannot define builtin, classifier, or prompt');
  }
  return policy;
}

/** Validate and apply the same defaults as the Python declaration models. */
export function validateServicePolicies(value: unknown): ServicePoliciesConfig & {
  local_mode: LocalPolicyMode;
  native_mode: NativePolicyMode;
  attachments: ServicePolicyAttachment[];
} {
  const raw = assertObject(value, 'service policies');
  assertKeys(raw, new Set(['local_mode', 'native_mode', 'attachments', 'abac']), 'service policies');
  const local_mode = enumValue<LocalPolicyMode>(raw.local_mode, LOCAL_MODES, 'mirror', 'local_mode');
  const native_mode = enumValue<NativePolicyMode>(raw.native_mode, NATIVE_MODES, 'off', 'native_mode');
  const attachmentsValue = raw.attachments ?? [];
  if (!Array.isArray(attachmentsValue)) throw new TypeError('service policy attachments must be an array');
  const attachments = attachmentsValue.map((value) => {
    const attachment = assertObject(value, 'service policy attachment');
    assertKeys(attachment, new Set(['name', 'target_type', 'target', 'mode', 'policies']), 'service policy attachment');
    const target_type = enumValue<ServicePolicyTargetType>(attachment.target_type, TARGET_TYPES, 'mcp_service', 'service policy target_type');
    const mode = enumValue<ServicePolicyMode>(attachment.mode, MODES, 'enforce', 'service policy mode');
    const policiesValue = attachment.policies ?? [];
    if (!Array.isArray(policiesValue)) throw new TypeError('service policy attachment policies must be an array');
    return {
      name: stringValue(attachment.name, 'service policy attachment name'),
      target_type,
      target: stringValue(attachment.target, 'service policy target'),
      mode,
      policies: policiesValue.map(normalizePolicy),
    } satisfies ServicePolicyAttachment;
  });
  if (native_mode === 'required' && attachments.length === 0) throw new TypeError("native_mode='required' needs at least one attachment");
  let abac: AbacSelector | undefined;
  if (raw.abac !== undefined) {
    const abacRaw = assertObject(raw.abac, 'service policy abac');
    assertKeys(abacRaw, new Set(['tags']), 'service policy abac');
    const tagsRaw = assertObject(abacRaw.tags ?? {}, 'service policy abac tags');
    const tags: Record<string, string> = {};
    for (const [key, tag] of Object.entries(tagsRaw)) tags[stringValue(key, 'ABAC tag key')] = stringValue(tag, 'ABAC tag value');
    abac = { tags };
  }
  return { local_mode, native_mode, attachments, abac };
}

/** Return applicable policies in deterministic rank order. */
export function orderedPolicies(policies: readonly ServicePolicy[], phase: ServicePolicyPhase | string): ServicePolicy[] {
  const requested = enumValue<ServicePolicyPhase>(phase, PHASES, 'on_call', 'service policy event phase');
  const applicable = policies.filter((policy) => {
    const policyPhase = policy.phase ?? 'on_call';
    return requested === 'both' || policyPhase === requested || policyPhase === 'both';
  });
  return [...applicable].sort((left, right) => {
    const leftRank = left.rank ?? 100;
    const rightRank = right.rank ?? 100;
    return requested === 'on_result' ? rightRank - leftRank : leftRank - rightRank;
  });
}

function normalizeAction(value: ServicePolicyEvaluationResult): { action: ServicePolicyAction; reason: string | null } {
  const raw = typeof value === 'string' ? { action: value } : value;
  if (!raw) return { action: 'ALLOW', reason: null };
  const action = String(raw.action).toUpperCase() as ServicePolicyAction;
  if (!new Set(['ALLOW', 'ASK', 'DENY', 'UNAVAILABLE']).has(action)) throw new TypeError(`unknown policy action ${raw.action}`);
  return { action, reason: raw.reason ?? null };
}

/** Evaluate in rank order, stopping at the first ASK or DENY and failing closed. */
export function evaluateOrderedServicePolicies(
  policies: readonly ServicePolicy[],
  event: ServicePolicyEvent,
  options: { mode: ServicePolicyMode; adapter: 'local' | 'native'; evaluator?: ServicePolicyEvaluator },
): ServicePolicyDecision {
  const mode = options.mode;
  for (const policy of orderedPolicies(policies, event.phase)) {
    let result: { action: ServicePolicyAction; reason: string | null };
    try {
      result = normalizeAction(options.evaluator?.(policy, event));
    } catch (error) {
      result = {
        action: mode === 'dry_run' ? 'UNAVAILABLE' : 'DENY',
        reason: `policy evaluation unavailable (${String(error)})`,
      };
    }
    if (result.action === 'ASK' && (event.target_type !== 'mcp_service' || event.phase !== 'on_call')) {
      result = {
        action: mode === 'dry_run' ? 'UNAVAILABLE' : 'DENY',
        reason: result.reason ?? 'ASK is supported only for MCP on_call policies',
      };
    }
    if (result.action !== 'ALLOW') {
      return {
        action: result.action,
        reason: result.reason,
        policy_name: policy.name,
        phase: event.phase,
        rank: policy.rank ?? 100,
        mode,
        adapter: options.adapter,
      };
    }
  }
  return { action: 'ALLOW', reason: null, policy_name: null, phase: event.phase, rank: null, mode, adapter: options.adapter };
}
